services/provider/client/client.go

```diff
@@ -10,82 +10,85 @@ import (
 )
 
 type OCSProviderClient struct {
-	Client  pb.OCSProviderClient
-	timeout time.Duration
+	Client     pb.OCSProviderClient
+	clientConn *grpc.ClientConn
+	timeout    time.Duration
 }
 
-// NewProviderClient creates a client to talk to OCS provider server
-func NewProviderClient(cc *grpc.ClientConn, timeout time.Duration) *OCSProviderClient {
-	return &OCSProviderClient{Client: pb.NewOCSProviderClient(cc), timeout: timeout}
-}
-
-// NewGRPCConnection returns a grpc client connection which can be used to create the consumer client
-// Note: Close the connection after use
-func NewGRPCConnection(serverAddr string, timeout time.Duration) (*grpc.ClientConn, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+// NewProviderClient creates a client to talk to the external OCS storage provider server
+func NewProviderClient(ctx context.Context, serverAddr string, timeout time.Duration) (*OCSProviderClient, error) {
+	apiCtx, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 
 	opts := []grpc.DialOption{
 		grpc.WithInsecure(),
 		grpc.WithBlock(),
 	}
 
-	conn, err := grpc.DialContext(ctx, serverAddr, opts...)
+	conn, err := grpc.DialContext(apiCtx, serverAddr, opts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to dial: %v", err)
 	}
 
-	return conn, err
+	return &OCSProviderClient{
+		Client:     pb.NewOCSProviderClient(conn),
+		clientConn: conn,
+		timeout:    timeout}, nil
+}
+
+// Close closes the gRPC connection of the external OCS storage provider client
+func (cc *OCSProviderClient) Close() error {
+	return cc.clientConn.Close()
 }
 
 // OnboardConsumer to validate the consumer and create StorageConsumer
 // resource on the StorageProvider cluster
-func (cc *OCSProviderClient) OnboardConsumer(token, name, capacity string) (*pb.OnboardConsumerResponse, error) {
+func (cc *OCSProviderClient) OnboardConsumer(ctx context.Context, token, name, capacity string) (*pb.OnboardConsumerResponse, error) {
```

Comment:
Can we save the context also in the `OCSProviderClient` and use the same again and again. WDYT?

Reply:
context needs to pass from the outside as the actual context (not code object) of the operation is unknown to the client itself and can change between call sites.